File: packages/rag/src/retriever.ts
/**
 * Retriever: unified RAG retrieval with optional rerank.
 */
import type { DocumentChunk, SearchResult, RerankResult } from "./types.js";
import type { VectorStore } from "./vector-store.js";
import type { Reranker } from "./reranker.js";

export interface RetrieverOptions {
  topK?: number;
  reranker?: Reranker;
}

/**
 * Retrieve documents by embedding query; optionally rerank results.
 */
export class Retriever {
  constructor(
    private vectorStore: VectorStore,
    private embedFn: (text: string) => Promise<number[]>,
    private options: RetrieverOptions = {}
  ) {}

  async retrieve(query: string, topK?: number, filter?: Record<string, unknown>): Promise<SearchResult[] | RerankResult[]> {
    const k = topK ?? this.options.topK ?? 10;
    const embedding = await this.embedFn(query);
    const results = await this.vectorStore.search(embedding, k * 2, filter);

    if (this.options.reranker) {
      return this.options.reranker.rerank(query, results, k);
    }
    return results.slice(0, k);
  }
}

File: python/src/agent_framework/rag/retriever.py
"""Retriever: unified RAG retrieval with optional rerank."""

from typing import Awaitable, Callable
from agent_framework.rag.types import RerankResult, SearchResult
from agent_framework.rag.vector_store import VectorStore
from agent_framework.rag.reranker import Reranker


class Retriever:
    """Retrieve documents by embedding query; optionally rerank."""

    def __init__(
        self,
        vector_store: VectorStore,
        embed_fn: Callable[[str], Awaitable[list[float]]],
        top_k: int = 10,
        reranker: Reranker | None = None,
    ):
        self._vector_store = vector_store
        self._embed_fn = embed_fn
        self._top_k = top_k
        self._reranker = reranker

    async def retrieve(
        self,
        query: str,
        top_k: int | None = None,
        filter_: dict | None = None,
    ) -> list[SearchResult] | list[RerankResult]:
        k = top_k or self._top_k
        embedding = await self._embed_fn(query)
        results = await self._vector_store.search(embedding, k * 2, filter_)

        if self._reranker:
            return await self._reranker.rerank(query, results, k)
        return results[:k]
